onnx/defs/map/defs.cc

```diff
@@ -0,0 +1,402 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "onnx/defs/function.h"
+#include "onnx/defs/schema.h"
+
+#include <algorithm>
+#include <numeric>
+
+namespace ONNX_NAMESPACE {
+static std::vector<std::string> all_map_types() {
+    auto map_key_types = OpSchema::all_map_key_types();
+    auto map_value_types = OpSchema::all_tensor_sequence_types();
+    static std::vector<std::string> all_map_types;
+    for (auto key_type : map_key_types) {
+        for (auto value_type : map_value_types) {
+            std::string map_type = "map("+ key_type + ", " + value_type + ")";
+            all_map_types.emplace_back(map_type);
+        }
+    }
+    return all_map_types;
+}
+
+static const char* MapConstruct_ver18_doc = R"DOC(
+Constructs either an empty map of a certain type specified by the key_type and value_type attributes,
+or a map structure with specified 'keys' and 'values'
+'keys' and 'values' must have the same number of elements.
+All 'keys' must have the same data type.
+All 'values' must all be of the same type (tensor, sequence or map) and have the same data type.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MapConstruct,
+    18,
+    OpSchema()
+        .SetDoc(MapConstruct_ver18_doc)
+        .Attr(
+            "key_type",
+            "The type of the keys present in the map output.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "value_type",
+            "The type of the values pairs present in the map output.",
+            AttributeProto::TYPE_PROTO,
+            OPTIONAL_VALUE)
+        .Input(0, "keys", "Keys.", "T", OpSchema::Optional)
+        .Input(1, "values", "Sequence enclosing the values.", "S", OpSchema::Optional)
+        .Output(0, "map", "Map.", "M")
+        .TypeConstraint(
+            "T",
+            OpSchema::all_map_key_tensor_types(),
+            "Constrain input keys to integral and string tensor types.")
+        .TypeConstraint(
+            "S",
+            OpSchema::all_tensor_sequence_types(),
+            "Constrain input values to a sequence of any tensor type.")
+        .TypeConstraint(
+            "M",
+            all_map_types(),
+            "Constrain output to a map type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+
+          const size_t numInputs = ctx.getNumInputs();
+          const auto* key_proto = ctx.getAttribute("key_type");
+          const auto* value_proto = ctx.getAttribute("value_type");
```

Comment:
It's safer to use something like `bool key_input_present = ctx.hasInput(0);` for checks below.